### PR TITLE
Modifications to Rewrite logic to accomodate correct logic

### PR DESCRIPTION
**Pull request by bordoni**
_Ref:_ [C#129410](http://central.tri.be/issues/129410)
---

### CODE REVIEW

File: src/Tribe/Rewrite.php [hunk 1]
```diff
@@ -325,6 +325,46 @@ protected function replace_array_key( &$array, $search, $replace ) {
...
+	 * @return array  Final version correct variable unset and merged.
+	 */
+	protected function merge_canonical_query_vars( array $query_vars, $url = null ) {
+
+		/**
+		 * Creates a map for merging values into one canonical param.
+		 *
+		 * @since TBD
+		 *
+		 * @param array          $map_canonical_query_args Associative array following the format: `[ 'eventDate' => [ 'event-date', 'event_date', 'tribe-bar-date' ], ]`.
```
Comment (lucatume) [inline src/Tribe/Rewrite.php]: Wrap this line, too long.

File: src/Tribe/Rewrite.php [hunk 2]
```diff
@@ -412,6 +452,8 @@ public function get_canonical_url( $url, $force = false ) {
 			unset( $query_vars['paged'] );
 		}
 
+		// Makes sure params with same functionality end up on the correct key
```
Comment (lucatume) [inline src/Tribe/Rewrite.php]: Missing full stop at the end.

File: src/Tribe/Rewrite.php [hunk 3]
```diff
@@ -534,14 +543,133 @@ public function get_canonical_url( $url, $force = false ) {
...
-		if ( $found ) {
+		if ( $found_template ) {
 			// Since we're caching let's not cache unmatched rules to allow for their later, valid resolution.
 			$this->canonical_url_cache[ $url ] = $resolved;
 		}
 
 		return $resolved;
 	}
 
+	protected function resolve_link_template( $found_template, $matched_vars ) {
```
Comment (lucatume) [inline src/Tribe/Rewrite.php]: Missing doc block

File: src/Tribe/Rewrite.php [hunk 4]
```diff
@@ -534,14 +543,133 @@ public function get_canonical_url( $url, $force = false ) {
...
+		$replaced = str_replace( array_keys( $replace ), $replace, $found_template );
+
+		// Remove trailing chars.
+		$path     = rtrim( $replaced, '?$' );
+		$resolved = trailingslashit( home_url( $path ) );
+
+		return $resolved;
+	}
+
+	public function match_rule_vars( $matched_vars ) {
```
Comment (lucatume) [inline src/Tribe/Rewrite.php]: Missing doc block

File: src/Tribe/Rewrite.php [hunk 5]
```diff
@@ -923,6 +1051,8 @@ public function parse_request( $url, array $extra_query_vars = [], $force = fals
 		 */
 		$query_vars = apply_filters( 'tribe_rewrite_parse_query_vars', $query_vars, $extra_query_vars, $url );
 
+		$query_vars = $this->merge_canonical_query_vars( $query_vars, $url );
```
Comment (lucatume) [inline src/Tribe/Rewrite.php]: Should this happen **after** the `tribe_rewrite_parse_query_vars` or before?

File: src/Tribe/Rewrite.php [hunk 6]
```diff
@@ -325,6 +325,46 @@ protected function replace_array_key( &$array, $search, $replace ) {
...
+		 * @since TBD
+		 *
+		 * @param array          $map_canonical_query_args Associative array following the format: `[ 'eventDate' => [ 'event-date', 'event_date', 'tribe-bar-date' ], ]`.
+		 * @param string         $url                      The input URL to resolve to a canonical one.
+		 * @param Tribe__Rewrite $this                     This rewrite object.
+		 */
+		$map_canonical_query_args = (array) apply_filters( 'tribe_rewrite_canonical_query_args', [], $url, $this );
+
+		foreach ( $map_canonical_query_args as $canonical_key => $merge_keys ) {
+			foreach( $merge_keys as $merge_key ) {
```
Comment (tr1b0t) [inline src/Tribe/Rewrite.php]: :no_entry_sign: There must be a space between foreach and the first parenthesis '('

<sub>hash: 34ca69eedbfc5b745c1342dd95e52002</sub>

File: src/Tribe/Rewrite.php [hunk 7]
```diff
@@ -534,14 +543,133 @@ public function get_canonical_url( $url, $force = false ) {
...
+			}
+
+			if ( ! array_intersect( array_keys( $matched_vars ), $query_vars ) ) {
+				return '';
+			}
+
+			/*
+			 * We use `end` as, by default, the localized version of the slug in the current language will be at the
+			 * end of the array.
+			 * @todo here we should keep a map, that has to generated at permalink flush time, to map locales/slugs.
```
Comment (tr1b0t) [inline src/Tribe/Rewrite.php]: :warning: Comment refers to a TODO task "here we should keep a map, that has to generated at permalink flush time, to map locales/slugs"

<sub>hash: ba7fa373f0099f45548570f3a688ec80</sub>

File: src/Tribe/Rewrite.php [hunk 8]
```diff
@@ -534,14 +524,133 @@ public function get_canonical_url( $url, $force = false ) {
...
```
Comment (tr1b0t) [inline src/Tribe/Rewrite.php]: :warning: Comment refers to a TODO task "here we should keep a map, that has to generated at permalink flush time, to map locales/slugs"

<sub>hash: f3fa0b5a1459de790f74a0186dfea822</sub>

**Pull request closed without merging**